integration_tests/offboard_attitude.cpp

```diff
@@ -29,70 +35,161 @@ TEST_F(SitlTest, OffboardAttitudeRate)
+        {
+            auto yaw_deg = close_loop(0.0f,0.0f,-60.0f,telemetry);
+            offboard->set_attitude_rate({yaw_deg.roll, yaw_deg.pitch, yaw_deg.yaw, yaw_deg.thrust});
+            std::this_thread::sleep_for(std::chrono::milliseconds(5));
+        }
+        else
+        {
+            std::cout << "Yaw angle acheived" << std::endl;
+            std::cout<< "Roll: "<<telemetry->attitude_euler_angle().roll_deg << " Pitch: " <<telemetry->attitude_euler_angle().pitch_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
+            break;
+        }
+     }
+    std::cout<< "Roll: "<<telemetry->attitude_euler_angle().roll_deg << " Pitch: " <<telemetry->attitude_euler_angle().pitch_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
+
+    // Balancing
+    for (int i=0; i<1000; i++)
+    {
+         if (telemetry->attitude_euler_angle().yaw_deg <= 0.2f)
+         {
+             auto yaw_bal = close_loop(0.0f,0.0f,0.0f,telemetry);
+             offboard->set_attitude_rate({yaw_bal.roll, yaw_bal.pitch, yaw_bal.yaw, yaw_bal.thrust});
+             std::this_thread::sleep_for(std::chrono::milliseconds(10));
+         }
+         else
+         {
+             std::cout << "Balanced Yaw" << std::endl;
+             std::cout<< "Roll: "<<telemetry->attitude_euler_angle().roll_deg << " Pitch: " <<telemetry->attitude_euler_angle().pitch_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
+             break;
+         }
+    }
+    
+    // Stopping offboard mode and Landing.
     offboard_result = offboard->stop();
     EXPECT_EQ(offboard_result, Offboard::Result::SUCCESS);
     action_ret = action->land();
     std::cout << "Landed! " << std::endl;
+
     std::this_thread::sleep_for(std::chrono::seconds(3));
-    std::cout<< "Pitch: "<<telemetry->attitude_euler_angle().pitch_deg << " Roll: " <<telemetry->attitude_euler_angle().roll_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
+    std::cout << "Landed" << std::endl;
+    std::cout<< "Roll: "<<telemetry->attitude_euler_angle().roll_deg << " Pitch: " <<telemetry->attitude_euler_angle().pitch_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
     EXPECT_EQ(action_ret, Action::Result::SUCCESS);
 }
 
+//Creating a closing function:
+Rates close_loop(float roll_rate, float pitch_rate, float yaw_rate, std::shared_ptr<Telemetry> telemetry)
+{
+    double roll_control = 6.0 * (double) (roll_rate - telemetry->attitude_euler_angle().roll_deg);
+    double pitch_control = 8.0 * (double) (pitch_rate - telemetry->attitude_euler_angle().pitch_deg);
+    double yaw_control = 3.80 * (double) (yaw_rate - telemetry->attitude_euler_angle().yaw_deg);
+    double thrust_control = 0.1 * (double)(10.5f - telemetry->position().relative_altitude_m);
+    if (thrust_control < 0.1)
+    {
+        thrust_control = 0.35;
```

Comment:
Are you sure this is correct? This means `thrust_control` will jump from `0.1` to `0.35` when < `0.1`?

Reply:
That is to prevent against negative thrust. Because, the quad would then free fall.